hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/execution/traceability/HederaTracer.java

```diff
@@ -73,6 +85,34 @@ public void init(final MessageFrame initialFrame) {
         }
     }
 
+    @Override
+    public void finalizeOperation(final MessageFrame initialFrame) {
+        if (areActionSidecarsEnabled) {
+            // make sure only valid actions are left in allActions
+            final var validActions =
+                    allActions.stream().filter(SolidityAction::isValid).toList();
+
```

Comment:
not only that, but it's just cleaner that way, it turns out.  done.

Reply:
Hmm, the action is not in its final state after `trackNewAction`,  though --- `finalizeActionFor` alters the action again. We should do the validation after action is finalized

Reply:
good catch! will fix - still one-at-a-time though.

Reply:
fixed!

Reply:
Can't see an obvious reason why we wouldn't be able to add to the list only after the finalization in `finalizeActionFor()`, sounds more prudent as well.

Reply:
the one that does this is the "synthetic" one, right? what seems weird to me is that that's where that is done at all (but I didn't want to change that). 

but wait ... i thought you earlier suggested to _not_ validate in one place but instead to validate one-at-a-time as actions were created/updated (to avoid a list traversal)?

(w.r.t. removing invalid actions from all actions: though all actions might get large certainly we don't believe that invalid actions will be? though I do get that remove from an arraylist is `O(nm)` ...)

Reply:
> the one that does this is the "synthetic" one, right? what seems weird to me is that that's where that is done at all (but I didn't want to change that).

Correct.

> but wait ... i thought you earlier suggested to _not_ validate in one place but instead to validate one-at-a-time as actions were created/updated (to avoid a list traversal)?

I'm saying having all the add and validation operations done one at a time in one function. For later though if it makes sense with the ordering comment


> (w.r.t. removing invalid actions from all actions: though all actions might get large certainly we don't believe that invalid actions will be? though I do get that remove from an arraylist is `O(nm)` ...)

Exactly